brokers/artifacts/broker.go

```diff
@@ -58,122 +57,125 @@ func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) err
 	b.PubStarted()
 	b.PrintInfo("Starting plugin artifacts broker")
 
-	b.cleanupPluginsDirectory()
 	pluginMetas, err := utils.GetPluginMetas(pluginFQNs, defaultRegistry, b.ioUtils)
 	if err != nil {
 		return b.fail(fmt.Errorf("Failed to download plugin meta: %s", err))
 	}
-	b.PrintInfo("Downloading plugin extensions")
 
 	err = utils.ResolveRelativeExtensionPaths(pluginMetas, defaultRegistry)
 	if err != nil {
 		return b.fail(err)
 	}
 
-	for _, meta := range pluginMetas {
-		err = b.ProcessPlugin(meta)
+	requestedPlugins := convertMetasToPlugins(pluginMetas)
+
+	toInstall := b.SyncPlugins(requestedPlugins)
+
+	for _, plugin := range toInstall {
+		err = b.ProcessPlugin(&plugin)
 		if err != nil {
 			return b.fail(err)
 		}
 	}
 
+	err = b.writeInstalledPlugins(toInstall)
+	if err != nil {
+		b.PrintInfo("WARN: Failed to log installed plugins: %s", err)
+	}
+
 	b.PrintInfo("All plugin artifacts have been successfully downloaded")
 	b.PubDone("")
-
 	return nil
 }
 
-func (b *Broker) cleanupPluginsDirectory() {
-	b.PrintInfo("Cleaning /plugins dir")
-	files, err := b.ioUtils.GetFilesByGlob(filepath.Join("/plugins", "*"))
+// ProcessPlugin downloads all undownloaded plugin extensions and places the
+// relevant artifacts in their appropriate location in the /plugins directory.
+// If a plugin already has artifacts downloaded for a given extension, that
+// extension is skipped.
+func (b *Broker) ProcessPlugin(plugin *model.CachedPlugin) error {
+	workDir, err := b.ioUtils.TempDir("", "vscode-extension-broker")
 	if err != nil {
-		// Send log about clearing failure but proceed.
-		// We might want to change this behavior later
-		b.PrintInfo("WARN: failed to clear /plugins directory. Error: %s", err)
-		return
+		return err
 	}
 
-	for _, file := range files {
-		err = b.ioUtils.RemoveAll(file)
+	b.PrintInfo("Processing plugin %s", plugin.ID)
+	numExtensions := len(plugin.CachedExtensions)
+	extensionIdx := 0
+	for URL, path := range plugin.CachedExtensions {
+		extensionIdx = extensionIdx + 1
+		b.PrintInfo("  Installing plugin extension %d/%d", extensionIdx, numExtensions)
+		if path != "" {
+			// Plugin has previously been downloaded, nothing needs to be done.
```

Comment:
maybe it makes sense to print log that already downloaded archive is used

Reply:
:+1: